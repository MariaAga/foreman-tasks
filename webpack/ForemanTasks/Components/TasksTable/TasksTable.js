import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Table } from 'foremanReact/components/common/table';
import { STATUS } from 'foremanReact/constants';
import MessageBox from 'foremanReact/components/common/MessageBox';
import { translate as __ } from 'foremanReact/common/I18n';
import Pagination from 'foremanReact/components/Pagination/PaginationWrapper';
import { getURIQuery } from 'foremanReact/common/helpers';
import createTasksTableSchema from './TasksTableSchema';
import { updateURlQuery } from './TasksTableHelpers';

const TasksTable = ({
  getTableItems,
  error,
  status,
  results,
  history,
  itemCount,
  pagination,
}) => {
  const url = history.location.pathname + history.location.search;
  const uriQuery = getURIQuery(url);

  useEffect(() => {
    getTableItems(url);
  }, [history.location.search]);

  if (status === STATUS.ERROR) {
    return (
      <MessageBox
        key="tasks-table-error"
        icontype="error-circle-o"
        msg={__(`Could not receive data: ${error && error.message}`)}
      />
    );
  }

  if (status === STATUS.PENDING && results.length === 0) {
    return <div />;
  }

  if (results.length === 0) {
    return <span>{__('No Tasks')}</span>;
  }

  const changeTablePage = ({ page, perPage }) => {
    updateURlQuery(
      {
        page,
        per_page: perPage,
      },
      history
    );
  };

  const setSortHistory = (by, order) => {
    updateURlQuery({ sort_by: by, sort_order: order }, history);
  };

  return (
    <div className="tasks-table">
      <Table
        key="tasks-table"
        columns={createTasksTableSchema(
          setSortHistory,
          uriQuery.sort_by,
          uriQuery.sort_order
        )}
        rows={results}
      />
      <Pagination
        className="tasks-pagination"
        viewType="table"
        itemCount={itemCount}
        pagination={pagination}
        onChange={changeTablePage}
        dropdownButtonId="tasks-table-dropdown"
      />
    </div>
  );
};

TasksTable.propTypes = {
  results: PropTypes.array.isRequired,
  getTableItems: PropTypes.func.isRequired,
  status: PropTypes.oneOf(Object.keys(STATUS)),
  error: PropTypes.instanceOf(Error),
  itemCount: PropTypes.number.isRequired,
  pagination: PropTypes.shape({
    page: PropTypes.number,
    perPage: PropTypes.number,
  }),
  history: PropTypes.object.isRequired,
};

TasksTable.defaultProps = {
  status: STATUS.PENDING,
  error: null,
  pagination: {
    page: 1,
    perPage: 20,
  },
};

export default TasksTable;
