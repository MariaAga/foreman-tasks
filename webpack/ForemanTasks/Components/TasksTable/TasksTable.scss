.tasks-pagination {
  margin-top: -6px;
}

#tasks-table {
  margin-bottom: 70px;
}
