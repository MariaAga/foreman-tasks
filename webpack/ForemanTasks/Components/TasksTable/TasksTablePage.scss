.tasks-pagination {
  margin-top: -6px;
}
.tasks-table {
  margin-bottom: 70px;
}

.tasks-table-wrapper {
  .export-csv {
    float: right;
  }
  .search-bar {
    padding-bottom: 20px;
  }
}

.param-name {
  display: inline-block;
  width: 10em;
}

.task-details pre {
  white-space: pre;
  word-break: normal;
}

.tasks-time-row .time-label {
  float: left;
}
