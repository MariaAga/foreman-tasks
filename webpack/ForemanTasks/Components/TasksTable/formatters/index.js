export { default as selectionHeaderCellFormatter } from './selectionHeaderCellFormatter';
export { default as selectionCellFormatter } from './selectionCellFormatter';

export { actionNameCellFormatter } from './actionNameCellFormatter';
export { actionCellFormatter } from './actionCellFormatter';
export { dateCellFormmatter } from './dateCellFormmatter';
export { durationCellFormmatter } from './durationCellFormmatter';
