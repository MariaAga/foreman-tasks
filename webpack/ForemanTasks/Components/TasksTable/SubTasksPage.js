import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import TasksTablePage from './';

export const SubTasksPage = props => {
  const parentTaskID = props.match.params.id;
  const getBreadcrumbs = actionName => ({
    breadcrumbItems: [
      { caption: __('Tasks'), url: `/foreman_tasks/tasks` },
      {
        caption: actionName,
        url: `/foreman_tasks/tasks/${parentTaskID}`,
      },
      { caption: __('Sub tasks') },
    ],
  });
  return (
    <TasksTablePage
      getBreadcrumbs={getBreadcrumbs}
      parentTaskID={parentTaskID}
      {...props}
    />
  );
};

SubTasksPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.object,
  }),
};

SubTasksPage.defaultProps = {
  match: {
    params: {},
  },
};
