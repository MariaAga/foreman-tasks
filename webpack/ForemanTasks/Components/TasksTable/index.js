import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import TasksTable from './TasksTable';
import reducer from './TasksTableReducer';
import * as actions from './TasksTableActions';
import { selectQuery } from '../TasksDashboard/TasksDashboardSelectors';
import {
  selectStatus,
  selectError,
  selectResults,
  selectPagitation,
  selectItemCount,
  selectSort,
} from './TasksTableSelectors';

const mapStateToProps = state => ({
  status: selectStatus(state),
  error: selectError(state),
  sort: selectSort(state),
  results: selectResults(state),
  query: selectQuery(state),
  pagination: selectPagitation(state),
  itemCount: selectItemCount(state),
});
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export const reducers = { tasksTable: reducer };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TasksTable);
