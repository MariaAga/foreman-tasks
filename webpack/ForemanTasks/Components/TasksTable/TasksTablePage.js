import React from 'react';
import PropTypes from 'prop-types';
import { getURIsearch } from 'foremanReact/common/urlHelpers';
import { Spinner, Button } from 'patternfly-react';
import PageLayout from 'foremanReact/routes/common/PageLayout/PageLayout';
import { translate as __ } from 'foremanReact/common/I18n';
import { getURIQuery } from 'foremanReact/common/helpers';
import ExportButton from 'foremanReact/routes/common/PageLayout/components/ExportButton/ExportButton';
import { STATUS } from 'foremanReact/constants';
import TasksDashboard from '../TasksDashboard';
import TasksTable from './TasksTable';
import { resolveSearchQuery, addSearchToURL } from './TasksTableHelpers';
import { TASKS_SEARCH_PROPS } from './TasksTableConstants';
import { ActionConfirmation } from './Components/ActionConfirmation';
import './TasksTablePage.scss';

const TasksTablePage = ({ getBreadcrumbs, history, ...props }) => {
  const url = history.location.pathname + history.location.search;
  const uriQuery = getURIQuery(url);
  const onSearch = searchQuery => {
    resolveSearchQuery(searchQuery, history);
    props.getTableItems(url);
  };

  const getSelected = () => {
    const selected = props.results.filter(item =>
      props.selectedRows.includes(item.id)
    );
    return selected.map(item => ({
      name: item.action,
      id: item.id,
      isCancelleble: item.availableActions.cancellable,
    }));
  };

  return (
    <div className="tasks-table-wrapper">
      <ActionConfirmation
        showModal={props.isCancelAllModalOpen}
        closeModal={props.hideCancelAllModal}
        title={__('Cancel Selected Tasks')}
        message={__(
          `This will stop ${props.selectedRows.length} tasks, putting them in the "canceled" state.  Are you sure?`
        )}
        onClick={() => {
          props.cancelSelected(getSelected(), url);
          props.hideCancelAllModal();
        }}
        confirmAction={__('Yes')}
        abortAction={__('No')}
      />
      <PageLayout
        searchable
        searchProps={TASKS_SEARCH_PROPS}
        onSearch={onSearch}
        breadcrumbOptions={getBreadcrumbs(props.actionName)}
        toastNotifications="foreman-tasks-cancel"
        toolbarButtons={
          <React.Fragment>
            {props.status === STATUS.PENDING && <Spinner size="lg" loading />}
            <ExportButton
              url={addSearchToURL('/foreman_tasks/tasks.csv', uriQuery)}
              title={__('Export All')}
            />
            <Button
              disabled={props.selectedRows.length < 1}
              onClick={props.showCancelAllModal}
            >
              {__('Cancel Selected')}
            </Button>
          </React.Fragment>
        }
        searchQuery={getURIsearch()}
        beforeToolbarComponent={<TasksDashboard history={history} />}
      >
        <TasksTable history={history} {...props} />
      </PageLayout>
    </div>
  );
};

TasksTablePage.propTypes = {
  results: PropTypes.array.isRequired,
  getTableItems: PropTypes.func.isRequired,
  getBreadcrumbs: PropTypes.func.isRequired,
  actionName: PropTypes.string,
  isSubTask: PropTypes.bool,
  status: PropTypes.oneOf(Object.keys(STATUS)),
  history: PropTypes.object.isRequired,
  cancelSelected: PropTypes.func.isRequired,
  selectedRows: PropTypes.arrayOf(PropTypes.string),
  isCancelAllModalOpen: PropTypes.bool.isRequired,
  showCancelAllModal: PropTypes.func.isRequired,
  hideCancelAllModal: PropTypes.func.isRequired,
};

TasksTablePage.defaultProps = {
  actionName: '',
  isSubTask: false,
  status: STATUS.PENDING,
  selectedRows: [],
};

export default TasksTablePage;
