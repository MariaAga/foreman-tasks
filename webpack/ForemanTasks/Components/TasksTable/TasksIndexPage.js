import React from 'react';
import { translate as __ } from 'foremanReact/common/I18n';
import TasksTablePage from './';

export const TasksIndexPage = props => {
  const getBreadcrumbs = () => ({
    breadcrumbItems: [{ caption: __('Tasks'), url: `/foreman_tasks/tasks` }],
  });
  return <TasksTablePage getBreadcrumbs={getBreadcrumbs} {...props} />;
};
