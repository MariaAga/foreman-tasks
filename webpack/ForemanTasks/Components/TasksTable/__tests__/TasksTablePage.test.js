import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';
import TasksTablePage from '../TasksTablePage';
import { minProps } from './TasksTable.fixtures';

const fixtures = {
  'render with minimal props': minProps,

  'render with Breadcrubs': {
    ...minProps,
    getBreadcrumbs: () => ({
      breadcrumbItems: [
        { caption: 'Tasks', url: `/foreman_tasks/tasks` },
        {
          caption: 'action Name',
          url: `/foreman_tasks/tasks/someid`,
        },
        { caption: 'Sub tasks' },
      ],
    }),
  },
};

describe('TasksTablePage', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TasksTablePage, fixtures));
});
