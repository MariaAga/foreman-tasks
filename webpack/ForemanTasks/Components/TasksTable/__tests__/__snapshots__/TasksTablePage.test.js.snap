// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksTablePage rendering render with Breadcrubs 1`] = `
<div
  className="tasks-table-wrapper"
>
  <PageLayout
    beforeToolbarComponent={
      <Connect(TasksDashboard)
        history={
          Object {
            "location": Object {
              "search": "",
            },
          }
        }
      />
    }
    breadcrumbOptions={
      Object {
        "breadcrumbItems": Array [
          Object {
            "caption": "Tasks",
            "url": "/foreman_tasks/tasks",
          },
          Object {
            "caption": "action Name",
            "url": "/foreman_tasks/tasks/someid",
          },
          Object {
            "caption": "Sub tasks",
          },
        ],
      }
    }
    onSearch={[Function]}
    searchProps={
      Object {
        "autocomplete": Object {
          "id": "searchBar",
          "searchQuery": "",
          "url": "tasks/auto_complete_search",
          "useKeyShortcuts": true,
        },
        "bookmarks": Object {
          "canCreateBookmarks": true,
          "documentationUrl": "4.1.5Searching",
          "url": "/api/bookmarks",
        },
        "controller": "tasks",
      }
    }
    searchQuery="a=b"
    searchable={true}
    toastNotifications="foreman-tasks-cancel"
    toolbarButtons={
      <React.Fragment>
        <ExportButton
          url="/foreman_tasks/tasks.csv?include_permissions=true"
        />
      </React.Fragment>
    }
  >
    <TasksTable
      actionName=""
      cancelTaskAction={[MockFunction]}
      error={null}
      getTableItems={[MockFunction]}
      history={
        Object {
          "location": Object {
            "search": "",
          },
        }
      }
      isSubTask={false}
      itemCount={2}
      location={
        Object {
          "pathname": "",
        }
      }
      pagination={
        Object {
          "page": 1,
          "perPage": 10,
        }
      }
      results={
        Array [
          "a",
          "b",
        ]
      }
      sort={
        Object {
          "by": "q",
          "order": "w",
        }
      }
      status="RESOLVED"
    />
  </PageLayout>
</div>
`;

exports[`TasksTablePage rendering render with minimal props 1`] = `
<div
  className="tasks-table-wrapper"
>
  <PageLayout
    beforeToolbarComponent={
      <Connect(TasksDashboard)
        history={
          Object {
            "location": Object {
              "search": "",
            },
          }
        }
      />
    }
    onSearch={[Function]}
    searchProps={
      Object {
        "autocomplete": Object {
          "id": "searchBar",
          "searchQuery": "",
          "url": "tasks/auto_complete_search",
          "useKeyShortcuts": true,
        },
        "bookmarks": Object {
          "canCreateBookmarks": true,
          "documentationUrl": "4.1.5Searching",
          "url": "/api/bookmarks",
        },
        "controller": "tasks",
      }
    }
    searchQuery="a=b"
    searchable={true}
    toastNotifications="foreman-tasks-cancel"
    toolbarButtons={
      <React.Fragment>
        <ExportButton
          url="/foreman_tasks/tasks.csv?include_permissions=true"
        />
      </React.Fragment>
    }
  >
    <TasksTable
      actionName=""
      cancelTaskAction={[MockFunction]}
      error={null}
      getTableItems={[MockFunction]}
      history={
        Object {
          "location": Object {
            "search": "",
          },
        }
      }
      isSubTask={false}
      itemCount={2}
      location={
        Object {
          "pathname": "",
        }
      }
      pagination={
        Object {
          "page": 1,
          "perPage": 10,
        }
      }
      results={
        Array [
          "a",
          "b",
        ]
      }
      sort={
        Object {
          "by": "q",
          "order": "w",
        }
      }
      status="RESOLVED"
    />
  </PageLayout>
</div>
`;
